enterprise/coderd/prebuilds/reconcile.go

```diff
@@ -0,0 +1,511 @@
+	}
+}
+
+func (c *StoreReconciler) isStopped() bool {
+	return c.stopped.Load()
+}
+
+// ReconcileAll will attempt to resolve the desired vs actual state of all templates which have presets with prebuilds configured.
+//
+// NOTE:
+//
+// This function will kick of n provisioner jobs, based on the calculated state modifications.
+//
+// These provisioning jobs are fire-and-forget. We DO NOT wait for the prebuilt workspaces to complete their
+// provisioning. As a consequence, it's possible that another reconciliation run will occur, which will mean that
+// multiple preset versions could be reconciling at once. This may mean some temporary over-provisioning, but the
+// reconciliation loop will bring these resources back into their desired numbers in an EVENTUALLY-consistent way.
+//
+// For example: we could decide to provision 1 new instance in this reconciliation.
+// While that workspace is being provisioned, another template version is created which means this same preset will
+// be reconciled again, leading to another workspace being provisioned. Two workspace builds will be occurring
+// simultaneously for the same preset, but once both jobs have completed the reconciliation loop will notice the
+// extraneous instance and delete it.
+func (c *StoreReconciler) ReconcileAll(ctx context.Context) error {
+	logger := c.logger.With(slog.F("reconcile_context", "all"))
+
+	select {
+	case <-ctx.Done():
+		logger.Warn(context.Background(), "reconcile exiting prematurely; context done", slog.Error(ctx.Err()))
+		return nil
+	default:
+	}
+
+	logger.Debug(ctx, "starting reconciliation")
+
+	err := c.WithReconciliationLock(ctx, logger, func(ctx context.Context, db database.Store) error {
+		snapshot, err := c.SnapshotState(ctx, db)
+		if err != nil {
+			return xerrors.Errorf("determine current snapshot: %w", err)
+		}
+		if len(snapshot.Presets) == 0 {
+			logger.Debug(ctx, "no templates found with prebuilds configured")
+			return nil
+		}
+
+		var eg errgroup.Group
+		// Reconcile presets in parallel. Each preset in its own goroutine.
+		for _, preset := range snapshot.Presets {
+			ps, err := snapshot.FilterByPreset(preset.ID)
+			if err != nil {
+				logger.Warn(ctx, "failed to find preset snapshot", slog.Error(err), slog.F("preset_id", preset.ID.String()))
+				continue
+			}
+
+			if !preset.UsingActiveVersion && len(ps.Running) == 0 && len(ps.InProgress) == 0 {
+				logger.Debug(ctx, "skipping reconciliation for preset; inactive, no running prebuilds, and no in-progress operations",
+					slog.F("template_id", preset.TemplateID.String()), slog.F("template_name", preset.TemplateName),
+					slog.F("template_version_id", preset.TemplateVersionID.String()), slog.F("template_version_name", preset.TemplateVersionName),
+					slog.F("preset_id", preset.ID.String()), slog.F("preset_name", preset.Name))
+				continue
+			}
+
+			eg.Go(func() error {
+				// Pass outer context.
+				err = c.ReconcilePreset(ctx, *ps)
```

Comment:
seems strange to me that we take the snapshot under the ReconciliationLock transaction, but then don't use that transaction for the individual reconciliation of presets.

Reply:
Reconciliation of presets are independent. If reconciliation for one preset will fail - we don't want to rollback changes for another presets.

<hr/>

Few more thoughts:

We need `ReconciliationLock` only to prevent other coderd replicas to execute `reconciliation iterations` in parallel because we run `Reconciliation Loop` on all replicas.

So I guess we don't really need transaction - we just need distributed lock. So we can use `pg_advisory_lock` instead of `pg_advisory_xact_lock`?
But I see that we don't use `pg_advisory_lock` anywhere.

<hr/>

fyi: `SnapshotState` also creates transaction because it maybe called outside `ReconciliationLock`. But this tx will be ignored within reconciliation loop - because we don't support nested txs.